base/staticint.jl

```diff
@@ -0,0 +1,111 @@
+Base.Bool(x::StaticInt{0}) = false
+Base.Bool(x::StaticInt{1}) = true
+
+Base.BigInt(@nospecialize(x::StaticInt)) = BigInt(Int(x))
+Base.Integer(x::StaticInt) = x
+(::Type{T})(@nospecialize(x::StaticInt)) where {T<:Integer} = T(_dynamic_int(x))
+(::Type{T})(x::Int) where {T<:StaticInt} = StaticInt(x)
+Base.convert(::Type{StaticInt{N}}, ::StaticInt{N}) where {N} = StaticInt{N}()
+
+Base.promote_rule(@nospecialize(T1::Type{<:StaticInt}), ::Type{T2}) where {T2<:Number} = promote_type(Int, T2)
+Base.promote_rule(@nospecialize(T1::Type{<:StaticInt}), ::Type{T2}) where {T2<:AbstractIrrational} = promote_type(Int, T2)
+for (S, T) in [(:Complex, :Real), (:Rational, :Integer), (:(Base.TwicePrecision), :Any)]
+    @eval function Base.promote_rule(::Type{$S{T}}, @nospecialize(SI::Type{<:StaticInt})) where {T<:$T}
+        promote_type($S{T}, Int)
+    end
+end
+Base.promote_rule(::Type{Union{Nothing,Missing}}, @nospecialize(T::Type{<:StaticInt})) = Union{Nothing,Missing,Int}
+Base.promote_rule(::Type{T1}, @nospecialize(T2::Type{<:StaticInt})) where {T1>:Union{Missing,Nothing}} = promote_type(T1, Int)
+Base.promote_rule(::Type{T1}, @nospecialize(T2::Type{<:StaticInt})) where {T1>:Nothing} = promote_type(T1, Int)
+Base.promote_rule(::Type{T1}, @nospecialize(T2::Type{<:StaticInt})) where {T1>:Missing} = promote_type(T1, Int)
+for T in [:Bool, :Missing, :BigFloat, :BigInt, :Nothing, :Any]
+    # let S = :Any
+    @eval begin
+        Base.promote_rule(@nospecialize(S::Type{<:StaticInt}), ::Type{$T}) = promote_type(Int, $T)
+        Base.promote_rule(::Type{$T}, @nospecialize(S::Type{<:StaticInt})) = promote_type($T, Int)
+    end
+end
+Base.promote_rule(@nospecialize(T1::Type{<:StaticInt}), @nospecialize(T2::Type{<:StaticInt})) = Int
+
+Base.:(%)(@nospecialize(n::StaticInt), ::Type{Integer}) = Int(n)
+
+Base.eltype(@nospecialize(T::Type{<:StaticInt})) = Int
+Base.iszero(::Zero) = true
+Base.iszero(@nospecialize(x::StaticInt)) = false
+Base.isone(::One) = true
+Base.isone(@nospecialize(x::StaticInt)) = false
+Base.zero(@nospecialize(x::Type{<:StaticInt})) = Zero()
+Base.one(@nospecialize(x::Type{<:StaticInt})) = One()
+
+for T in [:Real, :Rational, :Integer]
+    @eval begin
+        @inline Base.:(+)(x::$T, y::Zero) = x
+        @inline Base.:(+)(x::$T, @nospecialize(y::StaticInt)) = x + Int(y)
+        @inline Base.:(+)(x::Zero, y::$T) = x
+        @inline Base.:(+)(@nospecialize(x::StaticInt), y::$T) = Int(x) + y
+        @inline Base.:(-)(x::$T, y::Zero) = x
+        @inline Base.:(-)(x::$T, @nospecialize(y::StaticInt)) = x - Int(y)
+        @inline Base.:(*)(x::$T, y::Zero) = Zero()
+        @inline Base.:(*)(x::$T, y::One) = x
+        @inline Base.:(*)(x::$T, @nospecialize(y::StaticInt)) = x * Int(y)
+        @inline Base.:(*)(x::Zero, y::$T) = Zero()
+        @inline Base.:(*)(x::One, y::$T) = y
+        @inline Base.:(*)(@nospecialize(x::StaticInt), y::$T) = Int(x) * y
+    end
+end
```

Comment:
Something similar to my other comment applies to this block of code: it seems many of the `Zero` and `One` specific definitions could be dropped?
I see there's the added complexity of wanting to `@nospecialize` the generic `StaticInt` arguments here - but wouldn't it be reasonable to have similar desires for `One` and `Zero`?

A related question: the current design returns `Zero` and `One` for mixed-operations where it is is possible to statically determine that the output will be a `Zero` or `One` (e.g., `(*)(x::Zero, y::$T)`). That's neat, definitely, but it does seem a little subtle. Is there compelling reasons for wanting this?

Reply:
Specializing on `Zero` and `One` allows us to avoid doing the operation so we do want a unique dispatch on those types.

The overall idea is to preserve statically known info wherever possible but drop it as soon as we know it can't propagate it so that we avoid specialization. Perhaps we should worry about type promotion here, but if possible we should try to propagate the information as far as possible.

Reply:
FWIW, I just checked what the `@code_llvm` of just doing it the plain way (i.e., not specializing to `Zero` etc.) for e.g. `(+)(::Int, ::Zero)` is: there's no difference between specializing it manually and letting the compiler do it.
IIUC, the optimization is done automatically since the static parameter `N` in `StaticInt{N}` is marked as `Core.Const` and LLVM optimizes e.g. `Core.Const(0)` away from there.

Reply:
Thanks! I've changed those lines accordingly.